RNWCPP/ReactUWP/Views/ScrollViewManager.cpp

```diff
@@ -28,6 +30,289 @@ enum class ScrollViewCommands
+
+protected:
+  double m_snapToInterval;
+  bool m_isSnapToAlignmentSet;
+  winrt::Primitives::SnapPointsAlignment m_snapToAlignment;
+
+private:
+   void topScroll(winrt::ScrollViewer& scrollViewer, int64_t tag, const char* event, double x, double y, double zoom);
+};
+
+ScrollViewShadowNode::ScrollViewShadowNode()
+{
+  m_snapToInterval = 0;
+  m_isSnapToAlignmentSet = false;
+  m_snapToAlignment = winrt::Primitives::SnapPointsAlignment::Near;
+}
+
+void ScrollViewShadowNode::dispatchCommand(int64_t commandId, const folly::dynamic& commandArgs)
+{
+  auto scrollViewer = GetView().as<winrt::ScrollViewer>();
+  if (scrollViewer == nullptr)
+    return;
+
+  switch (commandId)
+  {
+    case static_cast<int64_t>(ScrollViewCommands::ScrollTo) :
+    {
+      double x = commandArgs[0].asDouble();
+      double y = commandArgs[1].asDouble();
+      bool animated = commandArgs[2].asBool();
+      bool horiz = scrollViewer.HorizontalScrollMode() == winrt::ScrollMode::Auto;
+      if (horiz)
+      {
+        // snap the x coordinate
+        if (m_isSnapToAlignmentSet && m_snapToInterval > 0.0)
+        {
+          x = snap(x, m_snapToAlignment, m_snapToInterval, scrollViewer.ViewportWidth());
+        }
+        // ChangeView parameters: x, y, zoom, animated
+        scrollViewer.ChangeView(x, nullptr, nullptr, !animated/*disableAnimation*/);
+      }
+      else
+      {
+        // snap the y coordinate
+        if (m_isSnapToAlignmentSet && m_snapToInterval > 0.0)
+        {
+          y = snap(y, m_snapToAlignment, m_snapToInterval, scrollViewer.ViewportHeight());
+        }
+        // ChangeView parameters: x, y, zoom, animated
+        scrollViewer.ChangeView(nullptr, y, nullptr, !animated/*disableAnimation*/);
+      }
+      break;
+    }
+
+    case static_cast<int64_t>(ScrollViewCommands::ScrollToEnd) :
+    {
+      bool animated = commandArgs[0].asBool();
+      bool horiz = scrollViewer.HorizontalScrollMode() == winrt::ScrollMode::Auto;
+      if (horiz)
+        scrollViewer.ChangeView(scrollViewer.ScrollableWidth(), nullptr, nullptr, !animated/*disableAnimation*/);
+      else
+        scrollViewer.ChangeView(nullptr, scrollViewer.ScrollableHeight(), nullptr, !animated/*disableAnimation*/);
+      break;
+    }
+  }
+}
+
+void ScrollViewShadowNode::createView()
+{
+  Super::createView();
+
+  auto scrollViewer = GetView().as<winrt::ScrollViewer>();
+  ScrollViewManager* viewManager = static_cast<ScrollViewManager*>(GetViewManager());
+  int64_t tag = m_tag;
+
+  scrollViewer.ViewChanging([this, tag](const winrt::IInspectable& sender, const winrt::ScrollViewerViewChangingEventArgs& args)
+  {
+    winrt::ScrollViewer scrollViewer = sender.as<winrt::ScrollViewer>();
+    topScroll(scrollViewer, tag,
+      "topScroll",
+      args.NextView().HorizontalOffset(),
+      args.NextView().VerticalOffset(),
+      args.NextView().ZoomFactor());
+
+  });
+}
+ 
+
+void ScrollViewShadowNode::updateProperties(const folly::dynamic&& props)
```

Comment:
>ScrollViewShadowNode::updateProperties [](start = 5, length = 38)

Basic question about updateProperties:  Are default values set through updateProperties(), or is it up to the framework to set defaults for everything?

Reply:
values only come through updateProperties for things that are set.  For cases where xaml defaults differ from react, we usually set the default in createView right after creating the xaml control.  There is a good point though that when a property is set and then unset and its value is null, we need to be careful to go back to the react default in some cases and not always clear the xaml prop.  (Image stretch mode comes to mind as one of these)

Reply:
Thanks.  Looking at createView I don't see any defaults being set which is creating some mismatching behaviors.  For example, ScrollView.showsHorizontalScrollIndicator defaults to true, but ScrollViewer.HorizontalScrollBarVisibility defaults to false.

---
In reply to: [274295173](https://github.com/Microsoft/react-native-windows/pull/2263#discussion_r274295173) [](ancestors = 274295173)